### ISSUE TITLE
M500 doesn't always save properly when done midprint

### ISSUE THREAD

**Issue by buzztiaan** [MarlinFirmware/Marlin#416]
M92 E700 during print.
M500 during print.

M503 shows Esteps now being 851 (wtf)

Doesn't always happen, but i now need to doublecheck if the correct settings are saved, and resave them (often)

**Comment by Lukelectro** [comment]
I think the folowing way would work to save eeprom settings during print:

Make sure heaters are not currently heating (to prevent overheating) and nothing is moving
disable interupts
save settings to eeprom
re-enable interupts
carry on with what you where doing.

Maybe simply blocking M500 during print (make it return a message like "can't save to eeprom during print") is a better solution, because turning off the heaters and stopping movement during a print, even for a short while, will affect the print. (And keeping the heaters on while disabling temperature control seems a even worse idea)

**Comment by boelle** [comment]
are you up to doing the fix to get it to work?

**Comment by alexborro** [comment]
How long does it takes to save the eeprom?? I was thinking about some
miliseconds, nothing more.. I don't think miliseconds will mess heaters
and/or movement...

By the way, I have already saved the eeprom during printing and no problem
at all.

Cheers.

Alex.

2015-01-02 18:50 GMT-02:00 Bo Herrmannsen <notifications@github.com>:

> are you up to doing the fix to get it to work?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ErikZalm/Marlin/issues/416#issuecomment-68561025>.
>



-- 
*"Não é o mais forte da espécie que sobrevive, nem o mais inteligente. É
aquele que se adapta melhor as mudanças"  ( Charles Darwin )*

Alex Borro

**Comment by boelle** [comment]
so for a dummie this means that means we can save while printing without any issue? given it does not take more than a few milis to save

**Comment by Lukelectro** [comment]
3.4ms per byte, as per page 22 of the atmega 328 datasheet. That's probably a typical specification, I don't know what's worst case.

quotes from datasheet:

"
Table 5-1.
 3.4 ms Erase and Write in one operation (Atomic Operation)
 1.8 ms Erase Only
1.8 ms Write Only
"


"
The EEPROM Write Enable Signal EEPE is the write strobe to the EEPROM. When address
and data are correctly set up, the EEPE bit must be written to one to write the value into the
EEPROM. The EEMPE bit must be written to one before a logical one is written to EEPE, other-
wise no EEPROM write takes place. The following procedure should be followed when writing
the EEPROM (the order of steps 3 and 4 is not essential):
1. Wait until EEPE becomes zero.
2. Wait until SELFPRGEN in SPMCSR becomes zero.
3. Write new EEPROM address to EEAR (optional).
4. Write new EEPROM data to EEDR (optional).
5. Write a logical one to the EEMPE bit while writing a zero to EEPE in EECR.
6. Within four clock cycles after setting EEMPE, write a logical one to EEPE.
The EEPROM can not be programmed during a CPU write to the Flash memory. The software
must check that the Flash programming is completed before initiating a new EEPROM write.
Step 2 is only relevant if the software contains a Boot Loader allowing the CPU to program the
Flash. If the Flash is never being updated by the CPU, step 2 can be omitted. See “Boot Loader
Support – Read-While-Write Self-Programming, ATmega88 and ATmega168” on page 268 for
details about Boot programming.

**Caution: An interrupt between step 5 and step 6 will make the write cycle fail**, since the
EEPROM Master Write Enable will time-out. If an interrupt routine accessing the EEPROM is
interrupting another EEPROM access, the EEAR or EEDR Register will be modified, causing the
interrupted EEPROM access to fail. It is recommended to have the Global Interrupt Flag cleared
during all the steps to avoid these problems.
"

**Comment by alexborro** [comment]
I have just counted how many write command we have: about 35 commands.. If
the most variable are bytes, so it will take less than 150ms.. lets say
200ms at maximum.



2015-01-02 19:16 GMT-02:00 Lukelectro <notifications@github.com>:

> 3.4ms per byte, as per page 22 of the atmega 328 datasheet. That's
> probably a typical specification, I don't know what's worst case.
>
> quotes from datasheet:
>
> "
> Table 5-1.
> 3.4 ms Erase and Write in one operation (Atomic Operation)
> 1.8 ms Erase Only
> 1.8 ms Write Only
> "
>
> "
> The EEPROM Write Enable Signal EEPE is the write strobe to the EEPROM.
> When address
> and data are correctly set up, the EEPE bit must be written to one to
> write the value into the
> EEPROM. The EEMPE bit must be written to one before a logical one is
> written to EEPE, other-
> wise no EEPROM write takes place. The following procedure should be
> followed when writing
> the EEPROM (the order of steps 3 and 4 is not essential):
> 1. Wait until EEPE becomes zero.
> 2. Wait until SELFPRGEN in SPMCSR becomes zero.
> 3. Write new EEPROM address to EEAR (optional).
> 4. Write new EEPROM data to EEDR (optional).
> 5. Write a logical one to the EEMPE bit while writing a zero to EEPE in
> EECR.
> 6. Within four clock cycles after setting EEMPE, write a logical one to
> EEPE.
> The EEPROM can not be programmed during a CPU write to the Flash memory.
> The software
> must check that the Flash programming is completed before initiating a new
> EEPROM write.
> Step 2 is only relevant if the software contains a Boot Loader allowing
> the CPU to program the
> Flash. If the Flash is never being updated by the CPU, step 2 can be
> omitted. See “Boot Loader
> Support – Read-While-Write Self-Programming, ATmega88 and ATmega168” on
> page 268 for
> details about Boot programming.
>
> *Caution: An interrupt between step 5 and step 6 will make the write cycle
> fail*, since the
> EEPROM Master Write Enable will time-out. If an interrupt routine
> accessing the EEPROM is
> interrupting another EEPROM access, the EEAR or EEDR Register will be
> modified, causing the
> interrupted EEPROM access to fail. It is recommended to have the Global
> Interrupt Flag cleared
> during all the steps to avoid these problems.
> "
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ErikZalm/Marlin/issues/416#issuecomment-68563070>.
>



-- 
*"Não é o mais forte da espécie que sobrevive, nem o mais inteligente. É
aquele que se adapta melhor as mudanças"  ( Charles Darwin )*

Alex Borro

**Comment by gregrebholz** [comment]
I would guess most values are int (2 bytes) or floats (4 bytes).  The EEPROM version identifier is a 4 byte character string (3 characters and null).  Other arduino sizes are long (4 bytes), double (4 bytes; 8 bytes on the Due), or boolean (1 byte).  

Anyway, probably closer to 1/2 second, but the stepper motor code is entirely dependent on receiving regular interrupts trigged by the Arduino's internal timer (handled by ISR(TIMER1_COMPA_vect)) and it will not be able to generate the stepper pulses during this time with interrupts disabled.  Presumably the blocks in the move planner just don't get popped, so movement can resume when interrupts are enabled again, but you should at least still see a noticeable halt in movement and it may throw off position... I have no idea how a stepper motor would behave if the pulse train to the coils froze for a while.  @ErikZalm did say "We can not disable the stepper interrupt during movement" in the comments above, so I think allowing M500 during movement should get a lot of rigorous testing before any implementation.  In the meantime better to block M500 rather than write corrupted data to the EEPROM, which is what is happening today.

**Comment by thawkins** [comment]
Can you not cheat an insert an action into the planner queue to write the eeprom contents, that way it should be action-ed between movements. the action can just suspend the heater ISR, if and turn it back on when finished.

**Comment by alhirzel** [comment]
I am not 100% familiar with the code here but it is very important to know all sources of interrupts if we're going to do this... I've listed what I know about, any others maybe buried in libraries?

  * extruder driving in stepper.cpp:TIMER0_COMPA_vect
  * stepper driving in stepper.cpp:TIMER1_COMPA_vect
  * hotend/hotbed control, reading filament width in temperature.cpp:TIMER0_COMPB_vect
  * 4000ms watchdog timer in watchdog.cpp:WDT_vect

@thawkins - if I understand, you're proposing that we forecast times when the stepper and extruder drivers will be in motion for 500ms so we have a safe EEPROM window in which we can disable interrupts, kick the EEPROM then sei() and be on our way. We would leave the watchdog running and disable the temp control for this time. Sound about right?

**Comment by alhirzel** [comment]
If we decide to do that, we have no way of guaranteeing the storage will occur (imagine a particularly crafty set of gcode that never gives us the desired gap). How do we deal with that possibility, slow the print or error to the user?

**Comment by thawkins** [comment]
I was suggesting that the eprom write code is run off a handler that is
driven of an action inserted into the motion queue, in that way the write
can only occure between movements and not during a movement, as it will be
syncronised with the stream of movement actions.

**Comment by thawkins** [comment]
i was not suggesting to insert it into a suitable length move, I was
suggesting that it is inserted between moves, which should be atomic.

**Comment by alhirzel** [comment]
But if the print pauses for 500ms mid-print that may cause undesirable results e.g. with a hot hotend resting in one place, it could disturb surface finish on the piece.

**Comment by alhirzel** [comment]
Yep I think I understand now =]  Just trying to think of the consequences

**Comment by thawkins** [comment]
I agree that there is potential for a pause on the print and the
introduction of a "hicky", but probably no worse than USB buffering pauses
introduce,.

And if it impacts the print then people should not do it... Their choice.

**Comment by alhirzel** [comment]
We could split the EEPROM write up into multiple operations, e.g. write each multibyte value separately but atomically. This would decrease the delay to <<200ms and reduce the effect I'm worried about. Can also ship it with a disclaimer or leave it disabled by default or something like that.

Last worry is what if EEPROM write fails, any avenue of that compromising the print?

**Comment by gregrebholz** [comment]
It's an interesting idea.  We could add an attribute to the typedef block_t in planner.h to label the block as an eeprom-write instead of the assumed behavior as a motion block.  M500 would insert the block into the motion planner, and the interrrupt handler would have to check that attribute around stepper.cpp line 321... completely skipping the pulse generation code and instead disabling all interrupts, writing EEPROM, enabled interrupts and clearing the block out of the motion queue.

You're still messing around in an area of Marlin where millisecond timing is critical, and lingering in a block of code for >500ms that normally completes in only a few... but it might work.

**Comment by alhirzel** [comment]
It's like we're writing our own RTOS. ;P

**Comment by gregrebholz** [comment]
Probably not the best place to ask, but does anyone know why Marlin doesn't have a "settings" typedef so the eeprom read/write can be done algorithmically instead of one named variable at a time?  It would make it much more difficult to screw up the read/write process when adding new variables to the settings, and the current approach with ifdef all over the place makes it very difficult to understand what exactly exists in any given user's eeprom.  Anyway, if it was in a typedef we could much more easily break down the M500.  Every time the stepper ISR is called and finds it's working on the M500 motion block it could write a fixed number of bytes of the settings typedef, and if successful advance a "bytes written" attribute in the motion block.  When complete it would clear the block (writing the eeprom version header last to ensure its only valid after the whole thing has been written).  As it currently stands the "write eeprom" is like "write variable x at offset A", "write variable y at offset B", etc.  and breaking all that up into a series of smaller atomic writes would be a complete mess I think.

**Comment by alhirzel** [comment]
Related to the above comment, if we are going to do sub-value writes (e.g. byte-at-a-time rather than value-at-a-time) we can totally abandon atomicity if we have a redundant set of EEPROM data and a switch to go between them.

**Comment by gregrebholz** [comment]
Each write operation will still need to be "atomic" in the sense that an interrupt during the write sequence will cause corrupted values to be written... so the process has to be "disable interrupts, write, enable interrupts" regardless of how much is being written.

There's still a benefit to redundant sets in the event that the system halts before a complete set is written.   Still though, this is a pretty major change and we're supposed to be focusing on stability and bug fixing.  "M500 writes corrupted eeprom values during movement/heating" is a bug... if I have time this weekend I'll put together a pull request to disable it until a working implementation of this can be proposed.

**Comment by Lukelectro** [comment]
I know I'm new here, but I would agree with disabling M500 during movement/print. Very much better then writing corrupt values.

(It would be nice though if you could pause a print, edit values, resume print, and if found good, pause, save and resume.)

**Comment by boelle** [comment]
an idea.... do save at next layer change? but writes on hold until the next layer change comes up

or would that not work ?

**Comment by boelle** [comment]
if not i will vote for the simple block as @ErikZalm said about 2-3 years ago

"We should block M500 if the printer is moving."

**Comment by nophead** [comment]
Read this comment: https://github.com/ErikZalm/Marlin/issues/416#issuecomment-68563070

It looks like you only need to disable interrupts between step 5 and step 6, which is only a couple of instructions. That shouldn't affect anything. This is assuming the interrupts don't use the EEPROM. 

If that is all that is needed the change is trivial.

**Comment by boelle** [comment]
if you know where to change and in general know what you are doing then yes... but me a total noob i dont want to risk screwing  up

**Comment by nophead** [comment]
I wasn't suggesting you personally do it but this discussion has assumed it can't be done with affecting the print but if the interrupts don't read from EEPROM it looks like it can easily be done. Does anybody know otherwise?

**Comment by boelle** [comment]
and this is the last one before we can start testing the heck out of it in the dev branch.. oh well i guess i'm just impatient :-/

**Comment by nophead** [comment]
Looks like the code that does the writes is in the arduino library and the section marked as critical here: https://github.com/arduino/Arduino/blob/3a8ad75bcef5932cfc81c4746a87ddbdbd7e6402/build/macosx/dist/eeprom.h#L325 is probably where interrupts need to be disabled.

**Comment by alexborro** [comment]
I agree with Chris, I think we can first try the easy path.
Otherwise I think queueing M500 among movement command is the better
option.

Cheers.

Alex.
Em 04/01/2015 18:08, "Chris" <notifications@github.com> escreveu:

**Comment by gregrebholz** [comment]
Just came across this and wanted to put it here for posterity:  http://www.nongnu.org/avr-libc/user-manual/optimization.html  Apparently in avr-gcc the sei() and cli() functions for disabling and enabling interrupts are macros to assembly statements, and compiler optimizations are free to reorder them with the surrounding code.  This means the exact statement order in the C code is not preserved in the conversion to assembly/machine code, unless compiler optimizations are completely disabled (which would be bad in our case).  Anyone know a workaround offhand?

**Comment by gregrebholz** [comment]
Also @nophead, the recommendation at the bottom of that comment is "It is recommended to have the Global Interrupt Flag cleared during all the steps to avoid these problems."  So that would be disable before step 1 and enable after step 6, which I believe is common practice when writing to eeprom.  The only example they cite is an interrupt handler that accesses EEPROM, but there may be other ways to break an eeprom write.  In general the eeprom write operation is very timing critical, and needs to run uninterrupted.  

The code from the arduino library is particularly interesting however, as it already includes the "cli" that results in global interrupts being disabled... and I don't see the sei that would reenable them... not sure what's going on there... but we may also not have control over when interrupts are functioning and not.  If we write to eeprom using the arduino libraries they may mess with our interrupts on their own terms, affecting stepper/heater control without our direction.  This doesn't explain how corrupted values are being written when M500 is run during operation.

**Comment by nophead** [comment]
The way to avoid the compiler reordering it is to use assembler, which is presumable why it is already written in assembler.

The only time critical section is 5 and 6. The only reason for disabling interrupts for the rest of it is if the interrupts mess with the EEPROM registers. Do any of the interrupts read from EEPROM?

Since it already disables interrupts there are two mysteries: why it goes wrong in the first place and how interrupts get enabled again.

The status register is listed in the clobbers, so perhaps the compiler restores it. I am not familiar with AVR assembler so I will have to read up on it.

**Comment by gregrebholz** [comment]
My comment about the cli() and sei() macros was referring to the proposal that we disable interrupts while processing the "faux motion block" in the planner.  We wouldn't want to write that whole thing in assembly.  But I've changed my opinion on the importance of that level of timing after some more reading --

I think now that there are three sources for eeprom-vs-interrupt problems.  The first is as you mentioned, other code (executed by an interrupt handler) trying to use eeprom at the same time as your eeprom write.  I think we're ok there as none of the interrupt handlers I've seen in Marlin access eeprom for reading or writing.  A second source is that the AVR has a single charge pump for both eeprom and flash memory, so the datasheet mentions having to ensure your code isn't writing to flash when you ask it to write to eeprom.  I don't think Marlin flashes itself, but its worth a little searching.  And the third source is the "timing critical" portion that eeprom_write_byte is handling in your referenced code.  The datasheet says that EEWE must be written within 4 cycles after EEMWE is written, and that is only half a microsecond on an 8Mhz clock.  The cli/sbi/sbi assembly will accomplish that.

So, I don't have an answer to either of your two mysteries.  They are mysterious. :)  Marlin only calls sei() to enable interrupts during initial init, during a kill() to allow LCD updates to occur, and in the spiSend/spiRec routines of Sd2Card.cpp for printing from SD cards.  It seems other developers call sei() explicitly -- https://github.com/pololu/libpololu-avr/blob/master/test/3pi-test/pushbuttons.c  but it really seems like that should be happening in the Arduino library... though I suppose it would have to first test if interrupts were enabled, to know if they should be enabled by the end of the call.

**Comment by nophead** [comment]
When I disassemble ConfigurationStore.o compiled for AtMega1284P I find it actually calls __eewr_byte_m1284p, not eeprom_write_byte which is written in the source code. I can't find source for __eewr_byte_m1284p and I can't work out how it gets called instead of eeprom_write_byte.

**Comment by galexander1** [comment]
This is from /usr/lib/avr/include/avr/eeprom.h on my PC:

   #define eeprom_write_byte     _EEPROM_CONCAT2 (__eewr_byte, _EEPROM_SUFFIX)

Hope this helps.

**Comment by nophead** [comment]
Hmm, yes it seems to come from there under linux and not from the arduino distribution. Not sure where the source code is.

So it seems the code varies according to the device and the OS it was compiled on as I am sure on Windows it uses the arduino version. So maybe it works for some people and not others?

**Comment by nophead** [comment]
Yes that is what is on my RPI, which is the version I dissembled but on
Windows there is no /usr so it must come from
arduino-1.0.5\hardware\tools\avr\avr\include\avr\eeprom.h which has the
inline assembler version we have been discussing. That one disables
interrupts, so it is a mystery why it doesn't work.

Looks like the avr library version is different. Some source code here:
https://cells-source.cs.columbia.edu/plugins/gitiles/toolchain/avr-libc/+/master/avr-libc-1.7.1/libc/misc/eewr_byte.S,
not necessarily the same version but it doesn't mask interrupts if
__AVR_XMEGA__
is set. However that seems unlikely on a Reprap board.

**Comment by nophead** [comment]
@buzztiaan,
   Which version of Arduino are you using and which OS?

**Comment by stv0g** [comment]
The Arduino IDE seems to use a patched version of `include/avr/eeprom.h`
https://github.com/arduino/toolchain-avr/blob/master/avr-libc-patches/eeprom.h#L325.
This patch protects the mentioned transition between step 5-6 by clearing the global interrupt flag.

But also the upstream avr-libs seems to only proctect these two instructions:
http://svn.savannah.nongnu.org/viewvc/trunk/avr-libc/include/avr/eeprom.h?revision=2416&root=avr-libc&view=markup.

The current implementation is far from perfect. I'm now reimplementing it part by part:

  - use `eeprom_*_block` functions to get rid of the C-level loop which will cause locking overhead.
  - use `eeprom_update_*` functions for writing. This should be faster. Because most of the time only a few EEPROM values have changed. Instead of 35 writes, wi might only have to do 2-3.
  - Replace Configuration_Store subsystem by generic approach.

So I will start working on that :sake:

**Comment by boelle** [comment]
i just hit the assign key so others can see quickly that you work on it

**Comment by nophead** [comment]
@stv0g,
  So if both the Arduino routines and the avr routines mask interrupts, where is the bug?

  Your proposed changes seem to be optimisations but nothing to do with the bug?

**Comment by stv0g** [comment]
Good question :D

I guess caused by dispatching a lot of eeprom_write_byte() calls.
The EEPROM access itself isn't blocking the execution in any way.

But by repeatedly calling eeprom_write_byte() we will 'busy wait' in
```
while (!eeprom_isready()) { }
```

This won't block the interrupt handlers, but probably something else.

**Comment by nophead** [comment]
It will just block the foreground, which needs to wait anyway, so I don't see it as an issue. You only need non blocking if you want to do something else in the foreground at the same time as M500.

The bug is that people report M500 writes corrupt values when the interrupt is running. So far we have found no explanation for that, unless there are some versions of the arduino library or the avr lib that don't mask interrupts.

**Comment by Lukelectro** [comment]
You also get corrupt values if some bytes of a float or int get written and some don't. So there needs to be a mechanism to catch or prevent that. (Might be already there)

The reason to disable interrupts in the steps before 5 and 6 is that if  EEPE, SELFPRGEN, EEAR and/or EEDR are changed in an interrupt the write will also fail (or occur at the wrong adress with the wrong data, thats also a failure although something does get written succesfully). As long as those bits / registers are not changed during any interupts, disabling global interupts during steps 5 and 6 would be sufficient.

Links to the discussion are broken by the move so I might have missed something.

**Comment by nophead** [comment]
I think consensus was the interrupts don't access the EEPROM. It would be mad if they did and since you can run with EEPROM disabled I assume all the value are copied to ram variables and never used directly.

So it all looks like it should work. Waiting to hear back from the OP what version of the eeprom library he is using as the critical bit of code is not actually in Marlin.

**Comment by stv0g** [comment]
I checked the source: We've only a single Function which accesses the EEPROM.
Thats `Config_StroreSettings.h`.

In my current approach, we'll have a Class called `ConfigSetting` which handles reads and writes to the EEPROM and as well holds the default (factory) value.
The actual configuration variable is only referenced by a pointer.

To be on the safe side:
Once we schedule a M500, I'd like make an additional copy of all values in the RAM.
This block of data is then fixed, and can be written to EEPROM by `eeprom_write_block`.

There's maybe only on drawback on this approach:
We will need a little bit more RAM for storing the default and "shadow copy" values.
At the moment the factory values should be part of the code section (not RAM).


One question: how critical is the usage of RAM?

**Comment by nophead** [comment]
Why do you want to take a copy? Do you think the values being written are being changed by the interrupt handlers? If that was the case it would only be particular values that are corrupted. It would be a lot of ram to waste without knowing it was definitely needed.

I really think it is the wrong approach to rewrite the code before finding out what the actual bug is. It could be that the only bug is using old versions of the libraries.

A complete bodge would be to read each value after it is written and loop to write it again if it doesn't match. At least that doesn't waste any RAM.

**Comment by stv0g** [comment]
Ok, I agree

**Comment by nophead** [comment]
Have you managed to re-create the bug with your setup?

**Comment by alexborro** [comment]
+1 for re-scan the values to double check. If it don't match, write again.





2015-01-07 8:38 GMT-02:00 Chris <notifications@github.com>:

> Have you managed to re-create the bug with your setup?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/issues/416#issuecomment-69004918>
> .
>



-- 
*"Não é o mais forte da espécie que sobrevive, nem o mais inteligente. É
aquele que se adapta melhor as mudanças"  ( Charles Darwin )*

Alex Borro

**Comment by boelle** [comment]
+1 from me also...

might i add to buffer writes and do them at a time where its predicted to be best done?

**Comment by scotty1024** [comment]
Please keep in mind EEPROM has limited write endurance when doing the retry loop. If something silly is happening with the EEPROM mid-print due to interference from all the power switching for steppers, hot end, heated bed etc ... The repeat logic could shorten the life span of the EEPROM bytes by quite a bit if it is designed to be aggressive (allow lots of tries).

**Comment by buzztiaan** [comment]
alright, found it finally :cake: 

i used arduino 1.0.5 and marlin_v1 commit 590225be2594e9d107ada8c40430308f09070a2a

**Comment by boelle** [comment]
so what happens with latest commit only a few sec's old?

**Comment by alexborro** [comment]
@scotty1024 you are right, but I think such writing error is not often (or rare).. But we can add a write limit and then prompt an error message. Like, it can try 3 times and if error persist, issue a fault message.

**Comment by stv0g** [comment]
And my optimizations will reduce the writes to a minimum…

**Comment by nophead** [comment]
@buzztiaan,
  Which OS are you compiling on? If it is Linux then you also need to look at the AVR lib version as that is where the write to EEPROM comes from. Seems like the include path is set incorrectly by the IDE as it makes more sense to use the avr lib bundled with the IDE.

@scotty1024,
   If your stepper motors affect the EEPROM you have a hardware problem. An unstable 5V rail might be enough to corrupt writes.

**Comment by thawkins** [comment]
Another possible approach is to hook an event to the end of the job, or the
timer getting no further data from the link, and deffer the eeprom write to
after the current job finishes.

We would however have to make sure we ignored the m105 temp queries. So a
variable in the outer loop that updated a timestamp for all active commands
except those might be a good idea. Check for current time (second count)
being more than 30 above last timestamp'

**Comment by stv0g** [comment]
Or suspending the GCode parsed until all bytes of a value have been written?
But there's still the problem that settings might be changed via the LCD panel.

**Comment by galexander1** [comment]
I think we're going in circles.  We started with imagining a simple fix
(disabling interrupts), moved on to excessive fixes (restructuring the
event queue, restructuring eeprom writes), then it was discovered that
the interrupts *are* disabled (probably) and the bug shouldn't be
happening, but now we're back to imagining excessive fixes.

The bug shouldn't be happening because it's already been fixed in the avr
lib.  We need to find out if the fix didn't work, or if the problem is
caused by version discrepancies or what-have-you.  Or if there is an
alternative cause we have yet to discover.

In other words, we need to start at the beginning and find a concrete
scenario where the bug happens and find out why it is happening in that
case.

One thing that might help is if someone who has demonstrated the problem
would send us the build that generated the problem.  A tar or zip of the
"applet" subdirectory should do it.  Then we can disassemble that and
concretely know if interrupts are suspended or not in that case.

**Comment by boelle** [comment]
could be funny to know what @buzztiaan  found 4 hours ago too

i know its sounds rough, but these long novels can make it very tempting to start ignore things... ie you can get so lost in them that you say either there is a problem with marlin or not... 

but yes... since @buzztiaan have the problem why dont he submit a copy....

**Comment by stv0g** [comment]
> The bug shouldn't be happening because it's already been fixed in the avr lib.

Thats only partially true. The aver-libc fixes this only for single-byte writes.

The issue here was observed by altering the `axis_steps_per_unit` array which is declared in `planner.cpp`:

```
float axis_steps_per_unit[NUM_AXIS];
```

This means we are talking about writing a value which is about 4*4 = 16 Bytes in size.
This takes 16 * 3.5 = 56ms only for a single setting.

The only way I can explain this bug is:
A float gets altered during the EEPROM write:

  - Byte 1-2 written to EEPROM
  - Float is changed
  - Byte 3-4 written

So we have to protect single values (not whole arrays) which are larger than 1 Byte.
This can be done by:

  - Mutexes
  - Comparing the value after it has been written to the EEPROM
  - Make a "shadow" copy of the value. Use the copy for writing.

**Comment by stv0g** [comment]
Btw: why the hell are we using so much floating point operations in a 8-Bit micro controller?!

**Comment by galexander1** [comment]
The bug you are envisioning does not exist.

axis_steps_per_unit is not modified in an ISR.  QED.

**Comment by stv0g** [comment]
Damn you're right.
I grepped the whole Marlin source for modifications of `axis_steps_per_unit[]`.
It's only touched my M92 in the main loop, by the LCD code and the Configuration Store.

I suggest that we will first try to reproduce the issue before make more assumptions or ideas how to fix it.

**Comment by boelle** [comment]
i just asked the testers if they can confirm the issue is there

**Comment by arno-millenaar** [comment]
Tested, any sequence I try. I can not reproduce the problem.  Every time I change by G92 the E value during print, values will be reported back by M503 (with or without a M500 store command) as entered by G92 during print.

**Comment by arno-millenaar** [comment]
BTW: used a rather demanding circular print (rings) as it will be very demanding on command buffer, etc.  Also FYI, used SD card as input for gcode to drive printer, while using serial connection to change value by G92, etc

**Comment by scotty1024** [comment]
Since the EEPROM is embedded in the CPU die, failures to correctly program represent a serious degradation in the CPU chip.

My +1 is to fail any M500 with no retry, just so people know things are getting squirrley.

**Comment by boelle** [comment]
each cell in the EEPROM are good for about 10000 cycles, i guess an 328 is about the same as a 644 from atmel... but well reducing the number of writes are always good

**Comment by boelle** [comment]
1 of the testers just said he/she cant reproduce the issue mentioned here

**Comment by boelle** [comment]
from irc chat:

@EvdZ|Ultimaker> boelle_dk, Maybe add a st_synchronize() to the M500 function. This is a bit safer. Then there is no motion while writing to eeprom. (And no disturbances in the stepper code)

@EvdZ|Ultimaker> boelle_dk, It can be a timing issue. It is maybe 1/1000 change it will go wrong.

**Comment by Lukelectro** [comment]
Agreed with "reproduce and investigate before fixing".

However. Idea:The value written does not need to be modified in an 
interupt for a write to fail. Here is how:

Start EEPROM write, setup registers
disable interutps
write first byte
enable interupts
setup eeprom registers for second byte of the float
INTERUPT!!! New data has arrived on UART - Value being written is being 
changed
disable interutps
write 2nd byte
enable interupts
etc.

Would be a coincidence if that happens, but maybe with just the right 
timing...

Getting the timing right might be easyer with a proper debugger so you 
can set breakpoints... (not possible in arduino, but possible in avr 
studio using a JTAGice or AVR dragon)

Op 07-01-15 om 16:53 schreef Steffen Vogel:

**Comment by nophead** [comment]
How does data arriving at the UART change any of the values being written?
We are currently processing M500, anything arriving while we do that will
just be buffered, not processed.

**Comment by boelle** [comment]
timeplan: the testers team have been asked to see if they can reproduce this one, so far none have... BUT... will give it 2 weeks, if no one can reproduce and document how this issue will be closed with an comment that it can be reopened if any can reproduce and document it. we have to move on or we will just waste time that could be better spent on the other issues

**Comment by stv0g** [comment]
:+1: I'll try it too.

**Comment by boelle** [comment]
great... then we have 2 testing ATM

**Comment by alexborro** [comment]
I have tried writing to EEPROM while printing under heavy CPU load and it
was perfect.

My CoreXY bot was moving at 450mm/s which results in really small time
between interrupts. No problems at all.

Cheers.

Alex.

2015-01-07 21:51 GMT-02:00 Bo Herrmannsen <notifications@github.com>:

> great... then we have 2 testing ATM
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/issues/416#issuecomment-69113972>
> .
>



-- 
*"Não é o mais forte da espécie que sobrevive, nem o mais inteligente. É
aquele que se adapta melhor as mudanças"  ( Charles Darwin )*

Alex Borro

**Comment by boelle** [comment]
3.... hmm... should we still give it the 2 full weeks and close?

**Comment by avluis** [comment]
+1 Cannot reproduce here either.

**Comment by stv0g** [comment]
We already have 5 people which are not able to reproduce the issue: @avluis @alexborro @arno-millenaar @fmalpartida 

I'll try it in a minute.
I would declare this issue as __not verified__ and close it.
Or we could move it to a future milestone.

**Comment by daid** [comment]
My 2 cents. Could be a hardware glitch. We had to replace a few Arduino's at Ultimaker due to the EEPROM not working properly (settings got lost all the time)

(Note that it's most likely a rare glitch, as we've only had a few of them, and we sold lots of Ultimakers)

**Comment by boelle** [comment]
just move on to the other milestones... and let us thinker what to do here

**Comment by boelle** [comment]
created a new label called hard to verify work on it later.... put that one on here and removed it from milestone 1.... so yaaa milestone 1 passed on to the next one... :-D

**Comment by thawkins** [comment]
I wasnt able to contribute to milestone one, my test machine is nearing
completion now, hopefully this weekend, so i will probaly will be able to
start contributing in the later milestones. I also run a 3d printing news
aggregation site, called microfabfricator.com which we are adding bunch of
contributer written articles to, one of the articles i would like to add is
how to setup for contribution to Marlin, and how to run the workflow.

**Comment by boelle** [comment]
dont feel bad, my machine is also down, need new bearings and last bit of electronics... will be another month before i can get it running... 

food on table is more important :-)

this one will not be closed yet, but we know 5 have tried to reproduce the error and they could not... we move on as not to waste to much time

**Comment by thinkyhead** [comment]
Old enough to punt...

**Comment by buzztiaan** [comment]
lol again?
'this is too old'
'it still exists'
'but meh its too old'

wtf man :D

**Comment by thinkyhead** [comment]
Open a new issue. Alert the TESTERS topic. Reference this issue in your new issue. (It gets you to the top again!)

**Comment by daid** [comment]
Honestly, who cares? You should not do an M500 during printing. It's silly.

Hard to impossible to reproduce. Lot of effort to fix. Low to almost no practical impact on the user. I would close it and be done with it. Not worth the limited time of the developers that are working on stuff.

**Comment by thawkins** [comment]
Or just ignore the command if the system is printing and just respond with
a message of "system printing, save ignored"

**Comment by nophead** [comment]
I think we concluded M500 will work just fine regardless of interrupts if avrlib is recent enough. And I think nobody has managed to recreate the bug except @buzztiaan, who didn't answer my question here: https://github.com/MarlinFirmware/Marlin/issues/416#issuecomment-69013296.

**Comment by thinkyhead** [comment]
Yeah. Consider M500 during printing to be not a currently supported use-case, and for the time being it can be mentioned in the documentation that it's not guaranteed to work, due to timing issues perhaps.

Meanwhile, if you are able to experiment with your robot, which exhibits the problem, and to determine a way to fix it, please submit a PR.

As the codebase generally improves, with kinks worked out in the interrupt and serial communication code, etc., this problem may get sorted out. If some simple tweak, such as adding a delay, turns out to work around the issue on your system, that would be good for us to know, even if we can't understand why.

My pet obsessions with Marlin have been in UX and code cleanup, so not too deep with timing. I've cleaned up the EEPROM code some, and on its face that code is pretty simple. Some underlying conflict, perhaps with thermocouples or noisy I/O, could boondoggle otherwise good code. So what we ultimately need is a larger sample of users to test it, and I admit it's hard to find testers with enough stake in it to carry it out. To make that easier you could supply some procedure, or a piece of G-Code that demonstrates the issue, and which when run reveals the problem pretty consistently. Even in putting that together you may find some pattern that has been missed.

I understand the complaint that problems like this can linger without resolution. You go to the doctor and tell him _when I do this it hurts_, and all he does is tell you _don't do that, 1000 coins please_. Let's hope medicine never goes altogether open source... but the way this project seems to work best is when those who face a problem are able to find the solution. If code is not your thing, you may try to recruit from near and far, someone to assist you. But you can get far just messing around, time permitting.

**Comment by boelle** [comment]
i will close this one again

1) as the others say you should not save during print
2) if you have a recently updated IDE / avrlib

i will also lock it so that it can only be opend by collabs if there is a VERY good reason to do so
